File: ListadoPersonasYDepartamentos/Domain/UseCases/DefaultUseCasePersonas.cs
using Domain.DTOs;
using Domain.Entities;
using Domain.Interfaces;
using Domain.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Domain.UseCases
{
    public class DefaultUseCasePersonas : IUseCasePersonas
    {
        private readonly IRepositoryPersonas _repositoryPersonas;
        private readonly IRepositoryDepartamentos _repositoryDepartamentos;

        public DefaultUseCasePersonas(IRepositoryPersonas repositoryPersonas, IRepositoryDepartamentos repositoryDepartamentos)
        {
            _repositoryPersonas = repositoryPersonas;
            _repositoryDepartamentos = repositoryDepartamentos;
        }

        public List<PersonaConNombreDeDepartamentoDTO> getListaPersonasConDepartamento()
        {
            String nombreDepartamento = "";
            List<PersonaConNombreDeDepartamentoDTO> listaPersonasConNombreDepartamento = new List<PersonaConNombreDeDepartamentoDTO>();
            List<Departamento> listaDepartamentos = _repositoryDepartamentos.getListaDepartamentos().ToList();

            foreach (Persona persona in _repositoryPersonas.getListaPersonas())
            {
                nombreDepartamento = listaDepartamentos
                    .Where(departamento => departamento.id == persona.idDepartamento)
                    .First().nombre;

                listaPersonasConNombreDepartamento.Add(
                    new PersonaConNombreDeDepartamentoDTO(
                        persona.id,
                        persona.nombre,
                        persona.apellido,
                        persona.direccion,
                        persona.telefono,
                        persona.fechaNac,
                        persona.imagen,
                        nombreDepartamento
                    )
                );
            }

            return listaPersonasConNombreDepartamento;
        }

        public PersonaConNombreDeDepartamentoDTO GetDetallePersona(int id)
        {
            Persona persona = _repositoryPersonas.getPersonaById(id);

            if (persona == null)
                return null;

            Departamento departamento = _repositoryDepartamentos.getDepartamentoById(persona.idDepartamento);

            return new PersonaConNombreDeDepartamentoDTO(
                persona.id,
                persona.nombre,
                persona.apellido,
                persona.direccion,
                persona.telefono,
                persona.fechaNac,
                persona.imagen,
                departamento.nombre
            );
        }

        public PersonaConListaDeDepartamentosDTO GetPersonaConListaDepartamentos(int id)
        {
            Persona persona = _repositoryPersonas.getPersonaById(id);
            List<Departamento> departamentos = _repositoryDepartamentos.getListaDepartamentos().ToList();

            return new PersonaConListaDeDepartamentosDTO(persona, departamentos);
        }

        public PersonaConListaDeDepartamentosDTO GetPersonaParaCrear()
        {
            Persona personaVacia = new Persona();
            List<Departamento> departamentos = _repositoryDepartamentos.getListaDepartamentos().ToList();

            return new PersonaConListaDeDepartamentosDTO(personaVacia, departamentos);
        }



        public int CrearPersona(Persona persona)
        {
            return _repositoryPersonas.crearPersona(persona);
        }

        public int ActualizarPersona(int id, Persona persona)
        {
            return _repositoryPersonas.actualizarPersona(id, persona);
        }

        public int EliminarPersona(int id)
        {
            return _repositoryPersonas.eliminarPersona(id);
        }

        public List<Persona> getListaPersonas()
        {
            return _repositoryPersonas.getListaPersonas().ToList();
        }
    }
}

File: SGE/unidad14/ListadoPersonas/src/domain/useCases/DefaultUseCasePersonas.ts
import { IUseCasePersonas } from '../interfaces/use-cases/IUseCasePersonas';
import { IRepositoryPersonas } from '../interfaces/repositories/IRepositoryPersonas';
import { IRepositoryDepartamentos } from '../interfaces/repositories/IRepositoryDepartamentos';
import { Persona } from '../entities/Persona';
import { PersonaConNombreDeDepartamentoDTO } from '../dtos/PersonaConNombreDeDepartamentoDTO';

export class DefaultUseCasePersonas implements IUseCasePersonas {

  constructor(
    private repositoryPersonas: IRepositoryPersonas,
    private repositoryDepartamentos: IRepositoryDepartamentos
  ) {
  }  
  /**
   * Obtiene la lista de personas incluyendo el nombre del departamento
   */
  async getListaPersonasConDepartamento(): Promise<PersonaConNombreDeDepartamentoDTO[]> {
    const personas = await this.repositoryPersonas.getListaPersonas();
    const departamentos = await this.repositoryDepartamentos.getListaDepartamentos();

    return personas.map(p => {
      const dept = departamentos.find(d => d.ID === p.IDDepartamento);
      return {
        ID: p.ID,
        Nombre: p.Nombre,
        Apellidos: p.Apellidos,
        FechaNacimiento: p.FechaNacimiento,
        Direccion: p.Direccion,
        Telefono: p.Telefono,
        Foto: p.Foto,
        NombreDepartamento: dept?.Nombre || 'Sin departamento'
      };
    });
  }

  /**
   * Obtiene la lista completa de personas
   */
  async getListaPersonas(): Promise<Persona[]> {
    return await this.repositoryPersonas.getListaPersonas();
  }

  /**
   * Obtiene el detalle de una persona por ID, incluyendo nombre de departamento
   */
  async getDetallePersona(id: number): Promise<PersonaConNombreDeDepartamentoDTO | null> {
    const persona = await this.repositoryPersonas.getPersonaById(id);
    if (!persona) return null;

    const departamento = await this.repositoryDepartamentos.getDepartamentoById(persona.IDDepartamento);

    return {
      ID: persona.ID,
      Nombre: persona.Nombre,
      Apellidos: persona.Apellidos,
      FechaNacimiento: persona.FechaNacimiento,
      Direccion: persona.Direccion,
      Telefono: persona.Telefono,
      Foto: persona.Foto,
      NombreDepartamento: departamento?.Nombre || 'Sin departamento'
    };
  }

  async crearPersona(persona: Persona): Promise<number> {
    return await this.repositoryPersonas.crearPersona(persona);
  }

  async actualizarPersona(id: number, persona: Persona): Promise<number> {
    return await this.repositoryPersonas.actualizarPersona(id, persona);
  }

  async eliminarPersona(id: number): Promise<number> {
    return await this.repositoryPersonas.eliminarPersona(id);
  }
}
